Stats: Date picker - create the remaining shortcuts

annacmc opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #81938

## Proposed Changes

* This PR begins the process of adding the remaining shortcuts, as shown here:
![image](https://github.com/Automattic/wp-calypso/assets/30754158/a94f0a3e-3ca9-4a7a-bbe4-acb26f99f356)


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Open Calypso Live link and navigate to Stats page.
* Apply feature flag: `?flags=stats/date-control`
* Open the new dropdown and confirm the new shortcuts are visible.
* Confirm that clicking the shortcuts updates the dates in the start and end input fields.

Note: This doesn't cover any styling of the control. That will be handled in follow-up PRs. Likewise the behaviour of the "All Time" and "Custom Range" shortcuts is not yet hammered out.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

annacmc committed:
add working shortcuts to list

a8ck3n commented on client/my-sites/stats/stats-date-control/stats-date-control-picker.tsx:
```diff
@@ -7,18 +7,67 @@ import DateControlPickerShortcuts from './stats-date-control-picker-shortcuts';
+		const href = `${ url }${ query }`;
+
+		page( href );
+	};
```
Instead of updating the page URL, I think we just want to update the text input field. So something like this maybe:

```
const refreshStartDate = ( daysToSubtract: number ) => {
	const newStartDate = calculateQueryDate( daysToSubtract );
	changeStartDate( newStartDate );
}
```

In that way we prevent changes from being applied unless the "Apply" button is clicked.

a8ck3n commented on client/my-sites/stats/stats-date-control/stats-date-control-picker.tsx:
```diff
@@ -7,18 +7,67 @@ import DateControlPickerShortcuts from './stats-date-control-picker-shortcuts';
-			},
+			id: 'today',
+			label: 'Today',
+			onClick: () => redirectToQueryDate( 0 ),
```
I think we should avoid using a function pointer here and just past the offsets. The way I understand it, selecting a shortcut updates the start and end input fields so we should probably provide an offset and a range. As an example, Today might look like this:

```
{
  id: 'today',
  label: 'Today',
  offset: 0, // offset from current date
  range: 30, // as in 30 days
}
```

Then we could pass a generic callback to the `DateControlPickerShortcuts` component that handled updating the input fields based on the offset/range values. Something like this:

```
  <DateControlPickerShortcuts shortcutList={ shortcutList, handleShortcutApplied } />
```

Then we'd do the updating of fields in that callback.

```
const handleShortcutApplied = ( daysOffset: number, range: number ) => {
	const newStartDate = calculateQueryDate( daysOffset );
	changeStartDate( newStartDate );
	const newEndDate = 0; // calc the necessary date value
	changeEndDate( newEndDate );
}
```

This way we avoid breaking the encapsulation of the control (via the function pointers in `props.shortcuts`) and only update the caller when the Apply button is clicked.

a8ck3n commented on client/my-sites/stats/stats-date-control/stats-date-control-picker.tsx:
```diff
@@ -7,18 +7,67 @@ import DateControlPickerShortcuts from './stats-date-control-picker-shortcuts';
+			id: 'custom-range',
+			label: 'Custom Range',
+			onClick: () => {}, // eslint-disable-line @typescript-eslint/no-empty-function
+			// Keep this empty, since the custom range is handled by the DateControlPickerDate component.
```
I'm not sure "Custom Range" should be something the user selects. I was thinking this would be set by the control when the current range didn't match any of the shortcuts. Or maybe it just auto-focuses the first input field when selected? We'll probably need more input from the team on this one.

a8ck3n reviewed (commented):
This is a solid start! I was tinkering with this as well so I shared a bit of what I was playing with as comments below.

a8ck3n committed:
DateControlPickerShortcuts uses shared callback.

Instead of having a callback defined in the shortcut itself which could potentially break encapsulation.

a8ck3n committed:
Update shortcut handling.

Shortcuts now specify an offset and a range which we use to determine the new start and end dates for the input fields.

a8ck3n committed:
Update DateControlPickerShortcut type definition.

a8ck3n committed:
Expose DateControlPickerShortcut type.

And use it in the callback definition instead of “any”.

a8ck3n committed:
Update DateControlPickerShortcutsProps.

Provide a specific function signature instead of falling back to “any”.

a8ck3n committed:
Clean up date formatting.

a8ck3n committed:
Reduce duplicated code.

Move the date math and formatting into local helper functions.

a8ck3n committed:
Temporary inline CSS.

Will need to remove this once we’ve styled the popover but for now, we need to be able to see the dates when testing things.

a8ck3n committed:
Move shortcuts up a level.

The shortcut list should come from the caller so we’re moving it outside the DateControlPicker component.

annacmc reviewed (commented):
This is looking great, I especially like having the shortcut list moved out of the component into a prop. One thing I'm curious about is how All Time views are going to be working here. Do we plan to get the range of days that the blog has been alive to replace the 400 range?

I feel like this is a good start though for a component that can be iterated on in follow up PR's so LGTM 👍  (however, because I created this PR I cannot approve it 🙈 )

grzegorz-cp reviewed (approved):
Looks good.
Let's remove the `custom range` in the next PR and consider adding an option to update time.

grzegorz-cp committed:
Merge conflicts
